File: web/src/lib/features.ts
/**
 * Canonical landmark -> model-input transform. TypeScript port.
 *
 * ==========================================================================
 *  THIS IS A PORT OF ml/src/features.py. THE TWO MUST AGREE EXACTLY.
 *  Verified by src/lib/features.test.ts against the same golden fixtures the
 *  Python test uses. If they drift, the model trains fine and then predicts
 *  garbage in the browser with NO error message anywhere.
 *  See docs/feature-spec.md before changing anything here.
 * ==========================================================================
 *
 * Why hands and body are normalized separately: a sign is defined by WHERE the
 * hand is and WHAT SHAPE it makes. Pose is normalized against the shoulders
 * (and includes both wrists, so it encodes location); hands are normalized
 * against their own wrist (so they encode handshape independent of arm
 * position). Both survive, and the result is invariant to camera distance.
 */

// --------------------------------------------------------------------------
// Landmark indices
// --------------------------------------------------------------------------

/**
 * 13 upper-body points kept from the 33 MediaPipe Pose returns. Legs and fine
 * facial landmarks carry no signing information and only add noise.
 */
export const POSE_INDICES = [0, 2, 5, 7, 8, 11, 12, 13, 14, 15, 16, 23, 24] as const;

const RAW_LEFT_SHOULDER = 11;
const RAW_RIGHT_SHOULDER = 12;

const HAND_WRIST = 0;
const HAND_MCP_INDICES = [5, 9, 13, 17] as const;

// --------------------------------------------------------------------------
// Tensor shape contract
// --------------------------------------------------------------------------

export const NUM_POSE = POSE_INDICES.length; // 13
export const NUM_HAND = 21;

export const POSE_DIMS = NUM_POSE * 3; // 39
export const HAND_DIMS = NUM_HAND * 3; // 63
export const BASE_DIMS = POSE_DIMS + 2 * HAND_DIMS; // 165
export const PRESENCE_DIMS = 2;
export const FEATURE_DIMS = BASE_DIMS * 2 + PRESENCE_DIMS; // 332

export const SEQUENCE_LENGTH = 64;

const EPSILON = 1e-6;

// --------------------------------------------------------------------------
// Types - the wire format, identical to the API request body
// --------------------------------------------------------------------------

export type LandmarkList = number[][] | null;

export interface Frame {
  pose?: LandmarkList;
  left_hand?: LandmarkList;
  right_hand?: LandmarkList;
}

// --------------------------------------------------------------------------
// Normalization
// --------------------------------------------------------------------------

/** (33, 3) raw pose -> (13, 3) shoulder-normalized, flattened to 39 numbers. */
export function normalizePose(pose33: number[][]): Float64Array {
  const left = pose33[RAW_LEFT_SHOULDER];
  const right = pose33[RAW_RIGHT_SHOULDER];

  const ox = (left[0] + right[0]) / 2;
  const oy = (left[1] + right[1]) / 2;
  const oz = (left[2] + right[2]) / 2;

  const dx = left[0] - right[0];
  const dy = left[1] - right[1];
  const dz = left[2] - right[2];
  let scale = Math.sqrt(dx * dx + dy * dy + dz * dz);
  if (scale < EPSILON) scale = 1.0; // degenerate frame: signer turned sideways

  const out = new Float64Array(POSE_DIMS);
  for (let i = 0; i < NUM_POSE; i++) {
    const p = pose33[POSE_INDICES[i]];
    out[i * 3] = (p[0] - ox) / scale;
    out[i * 3 + 1] = (p[1] - oy) / scale;
    out[i * 3 + 2] = (p[2] - oz) / scale;
  }
  return out;
}

/** (21, 3) raw hand -> wrist-normalized, flattened to 63 numbers. */
export function normalizeHand(hand21: number[][]): Float64Array {
  const wrist = hand21[HAND_WRIST];

  let sum = 0;
  for (const index of HAND_MCP_INDICES) {
    const p = hand21[index];
    const cx = p[0] - wrist[0];
    const cy = p[1] - wrist[1];
    const cz = p[2] - wrist[2];
    sum += Math.sqrt(cx * cx + cy * cy + cz * cz);
  }
  let span = sum / HAND_MCP_INDICES.length;
  if (span < EPSILON) span = 1.0;

  const out = new Float64Array(HAND_DIMS);
  for (let i = 0; i < NUM_HAND; i++) {
    const p = hand21[i];
    out[i * 3] = (p[0] - wrist[0]) / span;
    out[i * 3 + 1] = (p[1] - wrist[1]) / span;
    out[i * 3 + 2] = (p[2] - wrist[2]) / span;
  }
  return out;
}

// --------------------------------------------------------------------------
// Assembly
// --------------------------------------------------------------------------

function checkShape(landmarks: LandmarkList | undefined, expected: number): number[][] | null {
  if (landmarks === null || landmarks === undefined) return null;
  if (landmarks.length !== expected) {
    throw new Error(`expected ${expected} landmarks, got ${landmarks.length}`);
  }
  return landmarks;
}

/** frames -> (N, BASE_DIMS + PRESENCE_DIMS), flat and row-major. */
export function buildBaseFeatures(frames: Frame[]): Float64Array {
  const rowWidth = BASE_DIMS + PRESENCE_DIMS;
  const out = new Float64Array(frames.length * rowWidth);

  // A dropped pose detection mid-clip is common. Holding the previous value
  // damages a motion model far less than injecting a block of zeros, which
  // would read as a violent jump.
  let lastPose: Float64Array | null = null;

  for (let i = 0; i < frames.length; i++) {
    const frame = frames[i];
    const base = i * rowWidth;

    const pose = checkShape(frame.pose, 33);
    if (pose !== null) lastPose = normalizePose(pose);
    if (lastPose !== null) out.set(lastPose, base);
    // else: leave zeros until the first successful detection

    const left = checkShape(frame.left_hand, NUM_HAND);
    if (left !== null) {
      out.set(normalizeHand(left), base + POSE_DIMS);
      out[base + BASE_DIMS] = 1.0;
    }

    const right = checkShape(frame.right_hand, NUM_HAND);
    if (right !== null) {
      out.set(normalizeHand(right), base + POSE_DIMS + HAND_DIMS);
      out[base + BASE_DIMS + 1] = 1.0;
    }
  }

  return out;
}

/**
 * Linearly resample (numFrames, width) to (targetLength, width) along time.
 * Matches numpy.interp, including its clamping behaviour at both endpoints.
 */
export function resampleSequence(
  seq: Float64Array,
  numFrames: number,
  width: number,
  targetLength: number = SEQUENCE_LENGTH,
): Float64Array {
  if (numFrames === 0) throw new Error("cannot resample an empty sequence");

  const out = new Float64Array(targetLength * width);

  if (numFrames === 1) {
    for (let t = 0; t < targetLength; t++) out.set(seq.subarray(0, width), t * width);
    return out;
  }

  for (let t = 0; t < targetLength; t++) {
    // Equivalent to np.linspace(0, numFrames - 1, targetLength)
    const position = (t * (numFrames - 1)) / (targetLength - 1);

    let lower = Math.floor(position);
    if (lower >= numFrames - 1) lower = numFrames - 2;
    const frac = position - lower;

    const a = lower * width;
    const b = (lower + 1) * width;
    const dst = t * width;

    for (let d = 0; d < width; d++) {
      out[dst + d] = seq[a + d] + frac * (seq[b + d] - seq[a + d]);
    }
  }

  return out;
}

/**
 * (T, BASE+PRESENCE) -> (T, FEATURE_DIMS).
 * Column order is [base | velocity | presence] and is part of the contract.
 */
export function addVelocity(seq: Float64Array, targetLength: number): Float32Array {
  const inWidth = BASE_DIMS + PRESENCE_DIMS;
  const out = new Float32Array(targetLength * FEATURE_DIMS);

  for (let t = 0; t < targetLength; t++) {
    const src = t * inWidth;
    const dst = t * FEATURE_DIMS;

    for (let d = 0; d < BASE_DIMS; d++) {
      out[dst + d] = seq[src + d];
      // Velocity on the first frame is zero.
      out[dst + BASE_DIMS + d] = t === 0 ? 0 : seq[src + d] - seq[src - inWidth + d];
    }

    out[dst + BASE_DIMS * 2] = seq[src + BASE_DIMS];
    out[dst + BASE_DIMS * 2 + 1] = seq[src + BASE_DIMS + 1];
  }

  return out;
}

/**
 * Top-level entry point: wire-format frames -> (targetLength, FEATURE_DIMS),
 * returned flat and row-major.
 *
 * The order is load-bearing and must match features.py:
 *   1. normalize + assemble   2. resample   3. append velocity
 *
 * Resampling BEFORE differencing means velocity is always measured over a
 * uniform time step, so a 15fps phone and a 60fps webcam produce comparable
 * numbers.
 */
export function extractFeatures(
  frames: Frame[],
  targetLength: number = SEQUENCE_LENGTH,
): Float32Array {
  if (frames.length < 2) {
    throw new Error("need at least 2 frames to extract features");
  }

  const base = buildBaseFeatures(frames);
  const resampled = resampleSequence(
    base,
    frames.length,
    BASE_DIMS + PRESENCE_DIMS,
    targetLength,
  );
  return addVelocity(resampled, targetLength);
}

File: ml/src/features.py
"""
Canonical landmark -> model-input transform.

============================================================================
  THIS FILE IS THE SINGLE SOURCE OF TRUTH.
  It is duplicated at:
      api/app/features.py      (literal copy)
      web/src/lib/features.ts  (line-by-line TypeScript port)
  If those drift from this file, the model trains fine and predicts garbage
  in the browser with NO error message. See ml/tests/test_golden.py.
============================================================================

Design note: why hands and body are normalized separately
---------------------------------------------------------
A sign is defined by *where* the hand is (forehead vs. chest) AND *what shape*
it makes. If we normalized everything against the hand's own wrist we would
destroy location; if we normalized everything against the shoulders we would
lose handshape detail at low resolution.

So we split the job:
  * POSE landmarks are normalized against the shoulders. Because the pose
    includes both wrists, this is what encodes WHERE the hands are.
  * HAND landmarks are normalized against their own wrist and hand span.
    This is what encodes the HANDSHAPE, independent of arm position.

The result is invariant to how far the signer sits from the camera and where
they stand in frame, while keeping both pieces of linguistic information.
"""

from __future__ import annotations

from typing import Any, Sequence

import numpy as np

# --------------------------------------------------------------------------
# Landmark indices
# --------------------------------------------------------------------------

# MediaPipe Pose returns 33 landmarks. We keep 13 upper-body points; legs and
# fine facial points carry no signing information and only add noise.
#   0 nose | 2 left eye | 5 right eye | 7 left ear | 8 right ear
#   11/12 shoulders | 13/14 elbows | 15/16 wrists | 23/24 hips
POSE_INDICES: tuple[int, ...] = (0, 2, 5, 7, 8, 11, 12, 13, 14, 15, 16, 23, 24)

# Index of the shoulders *within the raw 33-point array*.
RAW_LEFT_SHOULDER = 11
RAW_RIGHT_SHOULDER = 12

# Mirror pairs, expressed as positions *within the reduced 13-point array*.
# Used by the horizontal-flip augmentation. Nose (position 0) has no partner.
POSE_MIRROR_PAIRS: tuple[tuple[int, int], ...] = (
    (1, 2),    # eyes
    (3, 4),    # ears
    (5, 6),    # shoulders
    (7, 8),    # elbows
    (9, 10),   # wrists
    (11, 12),  # hips
)

# MediaPipe Hand returns 21 landmarks. Index 0 is the wrist; 5/9/13/17 are the
# knuckles (MCP joints) of index/middle/ring/pinky.
HAND_WRIST = 0
HAND_MCP_INDICES: tuple[int, ...] = (5, 9, 13, 17)

# --------------------------------------------------------------------------
# Tensor shape contract
# --------------------------------------------------------------------------

NUM_POSE = len(POSE_INDICES)   # 13
NUM_HAND = 21

POSE_DIMS = NUM_POSE * 3       # 39
HAND_DIMS = NUM_HAND * 3       # 63
BASE_DIMS = POSE_DIMS + 2 * HAND_DIMS  # 165  (pose + left hand + right hand)
PRESENCE_DIMS = 2              # was a left / right hand detected this frame
FEATURE_DIMS = BASE_DIMS * 2 + PRESENCE_DIMS  # 332  (base + velocity + presence)

SEQUENCE_LENGTH = 64           # every clip is resampled to exactly this many frames

EPSILON = 1e-6                 # guards against divide-by-zero on degenerate frames


# --------------------------------------------------------------------------
# Parsing
# --------------------------------------------------------------------------

def _landmarks_to_array(landmarks: Any, expected: int) -> np.ndarray | None:
    """Convert one frame's landmark list into an (expected, 3) array.

    Returns None when the detector found nothing, which the caller turns into
    a zero block plus a presence flag of 0.
    """
    if landmarks is None:
        return None
    arr = np.asarray(landmarks, dtype=np.float64)
    if arr.shape != (expected, 3):
        raise ValueError(f"expected landmark shape ({expected}, 3), got {arr.shape}")
    return arr


def parse_sequence(frames: Sequence[dict]) -> tuple[list, list, list]:
    """Split the wire format into three per-frame lists.

    Each frame is a dict shaped like:
        {"pose": [[x,y,z] * 33] | None,
         "left_hand":  [[x,y,z] * 21] | None,
         "right_hand": [[x,y,z] * 21] | None}
    """
    poses, lefts, rights = [], [], []
    for frame in frames:
        poses.append(_landmarks_to_array(frame.get("pose"), 33))
        lefts.append(_landmarks_to_array(frame.get("left_hand"), NUM_HAND))
        rights.append(_landmarks_to_array(frame.get("right_hand"), NUM_HAND))
    return poses, lefts, rights


# --------------------------------------------------------------------------
# Normalization
# --------------------------------------------------------------------------

def normalize_pose(pose33: np.ndarray) -> np.ndarray:
    """(33, 3) raw pose -> (13, 3) shoulder-normalized pose.

    Origin becomes the midpoint between the shoulders; one unit becomes the
    shoulder width. This is what makes the model indifferent to how far the
    signer sits from the camera.
    """
    left_shoulder = pose33[RAW_LEFT_SHOULDER]
    right_shoulder = pose33[RAW_RIGHT_SHOULDER]

    origin = (left_shoulder + right_shoulder) / 2.0
    scale = float(np.linalg.norm(left_shoulder - right_shoulder))
    if scale < EPSILON:
        scale = 1.0  # degenerate frame (signer turned fully sideways)

    return (pose33[list(POSE_INDICES)] - origin) / scale


def normalize_hand(hand21: np.ndarray) -> np.ndarray:
    """(21, 3) raw hand -> (21, 3) wrist-normalized hand.

    Origin becomes the wrist; one unit becomes the mean wrist-to-knuckle
    distance. Strips away *where* the hand is so the model sees only its shape.
    Location is preserved separately by the pose wrists.
    """
    wrist = hand21[HAND_WRIST]
    centred = hand21 - wrist

    span = float(np.mean(np.linalg.norm(centred[list(HAND_MCP_INDICES)], axis=1)))
    if span < EPSILON:
        span = 1.0

    return centred / span


# --------------------------------------------------------------------------
# Assembly
# --------------------------------------------------------------------------

def build_base_features(frames: Sequence[dict]) -> np.ndarray:
    """Wire-format frames -> (num_frames, BASE_DIMS + PRESENCE_DIMS) array.

    When the pose is missing for a frame we carry forward the last good one.
    A dropped detection mid-clip is common and holding the previous value is
    far less damaging to a motion model than injecting a block of zeros.
    """
    poses, lefts, rights = parse_sequence(frames)
    num_frames = len(frames)

    out = np.zeros((num_frames, BASE_DIMS + PRESENCE_DIMS), dtype=np.float64)
    last_pose: np.ndarray | None = None

    for i in range(num_frames):
        if poses[i] is not None:
            last_pose = normalize_pose(poses[i])
        pose_block = last_pose if last_pose is not None else np.zeros((NUM_POSE, 3))

        left_present = lefts[i] is not None
        right_present = rights[i] is not None

        left_block = normalize_hand(lefts[i]) if left_present else np.zeros((NUM_HAND, 3))
        right_block = normalize_hand(rights[i]) if right_present else np.zeros((NUM_HAND, 3))

        out[i, :POSE_DIMS] = pose_block.reshape(-1)
        out[i, POSE_DIMS:POSE_DIMS + HAND_DIMS] = left_block.reshape(-1)
        out[i, POSE_DIMS + HAND_DIMS:BASE_DIMS] = right_block.reshape(-1)
        out[i, BASE_DIMS] = 1.0 if left_present else 0.0
        out[i, BASE_DIMS + 1] = 1.0 if right_present else 0.0

    return out


def resample_sequence(seq: np.ndarray, target_length: int = SEQUENCE_LENGTH) -> np.ndarray:
    """Linearly resample (N, D) to (target_length, D) along time.

    Signs recorded at different speeds or frame rates must land on the same
    tensor shape. Interpolating presence flags yields fractional values in
    [0, 1]; that is intentional and treated as a soft confidence.
    """
    num_frames, dims = seq.shape
    if num_frames == 0:
        raise ValueError("cannot resample an empty sequence")
    if num_frames == 1:
        return np.repeat(seq, target_length, axis=0)

    source_positions = np.arange(num_frames, dtype=np.float64)
    target_positions = np.linspace(0.0, num_frames - 1.0, target_length, dtype=np.float64)

    out = np.zeros((target_length, dims), dtype=np.float64)
    for d in range(dims):
        out[:, d] = np.interp(target_positions, source_positions, seq[:, d])
    return out


def add_velocity(seq: np.ndarray) -> np.ndarray:
    """(T, BASE+PRESENCE) -> (T, FEATURE_DIMS) by appending frame-to-frame deltas.

    Movement is a core parameter of a sign — "MOTHER" and "FATHER" share a
    handshape and differ mainly in where they land, while many pairs differ
    only in motion. Explicit velocity saves the model from having to infer it.

    Velocity is computed on the base dims only (not presence) and the first
    frame's velocity is zero. Column order is [base | velocity | presence].
    """
    base = seq[:, :BASE_DIMS]
    presence = seq[:, BASE_DIMS:]

    velocity = np.zeros_like(base)
    velocity[1:] = base[1:] - base[:-1]

    return np.concatenate([base, velocity, presence], axis=1)


def extract_features(
    frames: Sequence[dict],
    target_length: int = SEQUENCE_LENGTH,
) -> np.ndarray:
    """Top-level entry point: wire-format frames -> (target_length, FEATURE_DIMS).

    Order matters and must be identical in the TypeScript port:
        1. normalize + assemble   2. resample   3. append velocity

    Resampling BEFORE differencing means velocity is always measured over a
    uniform time step, so a 15fps phone and a 60fps webcam produce comparable
    numbers.
    """
    if len(frames) < 2:
        raise ValueError("need at least 2 frames to extract features")

    base = build_base_features(frames)
    resampled = resample_sequence(base, target_length)
    features = add_velocity(resampled)

    assert features.shape == (target_length, FEATURE_DIMS), (
        f"feature contract violated: got {features.shape}, "
        f"expected ({target_length}, {FEATURE_DIMS})"
    )
    return features.astype(np.float32)
